### ISSUE TITLE
Package-module namespace bug when proc has same name as module

### ISSUE THREAD

**Issue by c-blake** [nim-lang/Nim#11052]
Create a directory, say `"package"`.  Create a file within it, say `"module.nim"` with the single line defining an exported ` proc` of the same name as the module.  I.e. `package/module.nim` contains:
```Nim
proc module*() = echo "ran"
```

### Example

Given the package/module structure above, a new nim file in the same directory as `package` containing:
```nim
import package/module
module() #should succeed and does
module.module() #should succeed and does
when not defined(skipBug):
  module.module.module() #should fail but works
  module.module.module.module()  # should fail but works
```
has incorrect behavior.

With the same package/module structure as above, a nim file like the below also breaks which seems related (but could conceivably relate to additional problems):
```Nim
import macros, package/module
macro params(pro: typed{nkSym}): untyped =
  let impl = pro.symbol.getImpl
  if impl == nil: echo "broken: ", pro  #compile-time echo
params(module)  #breaks
params(module.module)  #breaks
params(module.module.module)  #breaks
```

### Current Output
```
ran
ran
ran
ran
```
and at *compile-time* for the second example:
```
broken: module
```

### Expected Output

The first example should not compile (unless `-d:skipBug` is passed).

The second example should output zero "broken:" lines at compile-time.

### Additional Information
You can add as many levels of `module.` as you want.  They all share the same incorrect behavior.  I only showed two extra levels.

Nim version
```
git hash: ce024a73bf9050a18b0cb7c6a985127c86ba14c7
```
but probably other versions break as well.

**Comment by Araq** [comment]
>  module.module.module() #should fail but works
>  module.module.module.module()  # should fail but works

A module contains itself so that you can write ``module.module()``. And then ``module.module.module()`` is also valid. Arguably. I need to think about it.

**Comment by c-blake** [comment]
Ok.  Well, there's still the `getImpl` returning `nil` on such `module.module` names aspect of the issue.

**Comment by pb-cdunn** [comment]
@c-blake , I was mainly worried about the case where a `proc` has the same name as the module in which it resides. Is that essentially the same problem?

**Comment by c-blake** [comment]
Yeah.  I changed the title of the issue to just the problem for your `cligen` app. If Araq wants a module to be its own package as well as also containing itself as a module then `module.module.module.module.module()` or as many times as you want is all "valid and as intended".  [ Such self-containment is unusual, though I suspect not unprecedented in namespace management.  Definitely arises in some fields of mathematics.  Personally it reminds me of https://en.wikipedia.org/wiki/Turtles_all_the_way_down but I can understand a certain appeal. ;-) ]

**Comment by disruptek** [comment]
> > module.module.module() #should fail but works
> > module.module.module.module()  # should fail but works
> 
> A module contains itself so that you can write `module.module()`. And then `module.module.module()` is also valid. Arguably. I need to think about it.

I doubt anyone really cares how it works, as long as it is consistent.  Enabling `module.module()` (or `module.module.module()`) makes the fact that one cannot `import module from module` even more surprising.

**Comment by c-blake** [comment]
There may also be an issue (separate or the same?) with regard to modules containing themselves and symbol resolution in that this also "fails" (@Araq may let us know if it is a "true failure" or "user misperception") with no "package" involvement:
```Nim
import macros

macro params(pro: typed{nkSym}): untyped =
  let impl = pro.symbol.getImpl
  if impl == nil: echo "broken: ", pro  #compile-time echo

when defined(workaround): #works as expected
  from strutils import nil
  proc align() = discard
  params(align)
else:
  import strutils #exports align(string,Natural,char):string
  proc align() = discard
  params(align.align) #type mismatch: got <proc() | proc(string,Natural,char): string>
```
If the above code is saved in a file named `align.nim` then it fails to compile with trimmed errors in the side-comment.

It works if it is saved in a file named `foo.nim` and the final line is appropriately changed to `params(foo.align)`.

It also always works if `-d:workaround` is given.

As further support for the importance of this issue, many people don't like to think of new names.  So, `module.proc` where `module` and `proc` are the same string seems a pretty common occurrence in the wild.  Indeed, there are 15 files in the current Nim repository that have the issue:

```
compiler/hlo.nim
compiler/liftlocals.nim
compiler/reorder.nim
compiler/semobjconstr.nim
lib/impure/re.nim
lib/posix/kqueue.nim
lib/pure/complex.nim
lib/pure/parsesql.nim
lib/pure/random.nim
lib/pure/unidecode/unidecode.nim
lib/std/editdistance.nim
lib/system/alloc.nim
nimsuggest/sexp.nim
tests/generics/mmodule_same_as_proc.nim
tools/fakedeps.nim
```
